src/compat.rs

```diff
@@ -3,55 +3,125 @@ use std::io::{Read, Write};
+// read waker slot for this, but any would do.
+//
+// Don't ever use this from multiple tasks at the same time!
+pub(crate) trait SetWaker {
+    fn set_waker(&self, waker: &task::Waker);
 }
 
-impl<S> HasContext for AllowStd<S> {
-    fn set_context(&mut self, context: (bool, *mut ())) {
-        self.context = context;
+impl<S> SetWaker for AllowStd<S> {
+    fn set_waker(&self, waker: &task::Waker) {
+        self.set_read_waker(waker);
     }
 }
 
-pub(crate) struct Guard<'a, S>(pub(crate) &'a mut WebSocket<AllowStd<S>>);
+impl<S> AllowStd<S> {
+    pub(crate) fn new(inner: S, waker: &task::Waker) -> Self {
+        let res = Self {
+            inner,
+            write_waker_proxy: Default::default(),
+            read_waker_proxy: Default::default(),
+        };
+
+        // Register the handshake waker as read waker for both proxies,
+        // see also the SetWaker trait.
+        res.write_waker_proxy.read_waker.register(waker);
+        res.read_waker_proxy.read_waker.register(waker);
+
+        res
+    }
+
+    // Set the read waker for our proxies: this is only supposed
+    // to be called by read (or handshake) operations, i.e. the
+    // Stream impl on the WebSocketStream.
+    //
+    // Reading can also cause writes to happen, e.g. in case of
+    // Message::Ping handling.
+    pub(crate) fn set_read_waker(&self, waker: &task::Waker) {
+        self.write_waker_proxy.read_waker.register(waker);
+        self.read_waker_proxy.read_waker.register(waker);
+    }
 
-impl<S> Drop for Guard<'_, S> {
-    fn drop(&mut self) {
-        trace!("{}:{} Guard.drop", file!(), line!());
-        (self.0).get_mut().context = (true, std::ptr::null_mut());
+    // Set the write waker for our proxies: this is only supposde
+    // to be called by write operations, i.e. the Sink impl on the
+    // WebSocketStream.
+    pub(crate) fn set_write_waker(&self, waker: &task::Waker) {
+        self.write_waker_proxy.write_waker.register(waker);
+        self.read_waker_proxy.write_waker.register(waker);
     }
 }
 
-// *mut () context is neither Send nor Sync
-unsafe impl<S: Send> Send for AllowStd<S> {}
-unsafe impl<S: Sync> Sync for AllowStd<S> {}
+// Proxy Waker that we pass to the internal AsyncRead/Write of the
+// stream underlying the websocket. We have two slots here for the
+// actual wakers to allow external read operations to trigger both
+// reads and writes, and the same for writes.
+#[derive(Debug, Default)]
+struct WakerProxy {
+    read_waker: task::AtomicWaker,
+    write_waker: task::AtomicWaker,
+}
+
+impl task::ArcWake for WakerProxy {
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        arc_self.read_waker.wake();
+        arc_self.write_waker.wake();
+    }
+}
 
 impl<S> AllowStd<S>
 where
     S: Unpin,
 {
-    fn with_context<F, R>(&mut self, f: F) -> Poll<std::io::Result<R>>
+    fn with_context<F, R>(&mut self, read: bool, f: F) -> Poll<std::io::Result<R>>
```

Comment:
What if we use an enum here so it is a little more explicit?

```rust
enum ContextWaker {
  Read,
  Write,
}

impl ContextWaker {
  pub fn waker<S>(&self, allow_std: &AllowStd<S>) -> &task::AtomicWaker {
    match self {
      Read => &allow_std.read_waker_proxy,
      Write => &allow_std.write_walker_proxy,
    }
  }
}
```

Reply:
Done (slightly different, the above also won't work because then the return value keeps `allow_std` borrowed and it can't be mutably borrowed anymore for the `Pin`)